// vscode-languageworker/lib/messages.js
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
'use strict';
/**
 * Message kinds.
 */
var Message;
(function (Message) {
    /**
     * A request message usually send from the client to the server
     */
    Message.Request = 'request';
    /**
     * A message response usually send from the server to the client
     */
    Message.Response = 'response';
    /**
     * A event message usually send asynchronously from the server to the client
     */
    Message.Event = 'event';
})(Message = exports.Message || (exports.Message = {}));
/**
 * Tests if the given message is a request message
 */
function isRequestMessage(message) {
    return message.type === Message.Request;
}
exports.isRequestMessage = isRequestMessage;
/**
 * Tests if the given message is a event message
 */
function isEventMessage(message) {
    return message.type === Message.Event;
}
exports.isEventMessage = isEventMessage;
/**
 * Tests if the given message is a response message
 */
function isReponseMessage(message) {
    return message.type === Message.Response;
}
exports.isReponseMessage = isReponseMessage;
function isString(str) {
    return typeof (str) === 'string' || str instanceof String;
}
exports.isString = isString;
function isResponse(value) {
    var candidate = value;
    return candidate && (candidate.success === true || candidate.success === false);
}
exports.isResponse = isResponse;
function isFailedResponse(value) {
    var candidate = value;
    return isResponse(value) && candidate.success === false && isString(candidate.message);
}
exports.isFailedResponse = isFailedResponse;
function isSuccessfulResponse(value) {
    var candidate = value;
    return isResponse(value) && candidate.success;
}
exports.isSuccessfulResponse = isSuccessfulResponse;
